### ISSUE TITLE
UserSession.authorize() 2nd param "response"

### ISSUE THREAD

**Issue by btpschroeder** [Esri/arcgis-rest-js#627]
Could somebody clarify what the required 2nd parameter for `UserSession.authorize()` is supposed to be? The documentation indicates it should be a `ServerResponse`, but I'm having trouble finding any further info.
Not passing it in gives me this (probably expected) error:

![image](https://user-images.githubusercontent.com/19781811/65646355-b0345000-dfaf-11e9-836f-069c315a325e.png)

Relevant part of docs: https://esri.github.io/arcgis-rest-js/api/auth/UserSession/#authorize

Thanks!

**Comment by jgravois** [comment]
a server response is indeed what's expected.

check out the Express demo for a working example: https://github.com/Esri/arcgis-rest-js/tree/master/demos/express

**Comment by patrickarlt** [comment]
It is expected to be (at least) a `http.ServerResponse` https://nodejs.org/api/http.html#http_class_http_serverresponse which might be wrapped in your framework like https://expressjs.com/en/4x/api.html#res in Express or https://github.com/hapijs/hapi/blob/master/API.md#request.raw in HAPI

Depending on your exact framework you need to pass in something. You can see how we use this in https://github.com/Esri/arcgis-rest-js/blob/master/packages/arcgis-rest-auth/src/UserSession.ts#L435-L441 (hint we just call `writeHead()` and `end()`).

@jgravois this needs more docs. I'm assigning to me.

**Comment by btpschroeder** [comment]
Thank you both! I'm guessing this method was written to be handled on the back end with the [traditional auth model](https://developers.arcgis.com/documentation/core-concepts/security-and-authentication/server-based-user-logins/), which our app doesn't use. This is all being done on the front end for various reasons I won't get into here 😁 

Really, I just need a wrapper to be able to make a call to `/authorize` from our UI -- I could just wing it with Axios or whatever, but the rest of our product uses this library so I'd like to be consistent if at all possible.

Basically:

1. Our front end makes a call to our back end. If back end doesn't have a refresh token, it'll return an error code (403 etc).
2. If our front end detects that error code, it'll make a call to https://something.arcgis.com/sharing/rest/oauth2/authorize?client_id=[id]&response_type=code&redirect_uri=[blahblah] to get the refresh token.
3. Our front end takes that refresh token and sends it to the back end to work some unrelated magic.

Assuming this all has to be done on the front end, do we have any options here as it pertains to this lib?

**Comment by patrickarlt** [comment]
> I'm guessing this method was written to be handled on the back end with the traditional auth model, which our app doesn't use. This is all being done on the front end for various reasons I won't get into here 😁

We absolutely need to get into this :) there are 2 Oauth flows that REST JS supports

1. Client side (no refresh tokens, Oauth "Implicit Flow")
2. Server side (refresh tokens, Oauth "Authorization Code Flow")

What you are trying to do is server side auth on the client which is REALLY bad for security.

Server side oauth normally works like this:

1. The server redirects the user to the ArcGIS authorization page `UserSession.authorize()`](https://github.com/Esri/arcgis-rest-js/blob/master/packages/arcgis-rest-auth/src/UserSession.ts#L426-L442) where the user enters their credentials.
2. The authorization page generates a authorization code (not a token) and redirects user back to your redirect URL.
3. Your redirect URL should handle exchanging the code for a token and refresh token [`UserSession.exchangeAuthorizationCode()`](https://github.com/Esri/arcgis-rest-js/blob/master/packages/arcgis-rest-auth/src/UserSession.ts#L450-L485)
3. Your server then has the refresh token to store securely.

Since a refresh token represents PERMANENT and UNREVOKABLE access to the users account you should NEVER store or use it client side where it could be compromised (hence why the refresh is delivered to the server). [auth0 has some good doc on this](https://auth0.com
/docs/tokens/refresh-token/current).

It isn't even possible to implement this on the client because the oauth endpoint will respond with and HTML page that the user has to fill out in order to generate the code and then redirects to a new page to generate the code at step 2.

---

If you want refresh tokens you need to get your own Node server running and do proper oAuth (sorry).  You can store the token (not the refresh token on the client) and if the token expires generate a new token on the server. You can see an example of that here https://github.com/Esri/arcgis-rest-js/blob/master/demos/webmap-checker-sapper/src/utils.js#L4-L33 and https://github.com/Esri/arcgis-rest-js/blob/master/demos/webmap-checker-sapper/src/routes/auth/exchange-token.js. If the token expires we can refresh server side and re-run anything on the client.

---

If you cannot run your own server you need to use the client side Oauth flows in https://github.com/Esri/arcgis-rest-js/tree/master/demos/oauth2-browser. Since you can freely send regular tokens your backend making your workflow look like this:

1. Frontend checks for a cookie with a token
2. If the cookie exists send the token to the backend to do the work
3. If a cookie doesn't exist or the token has expired ask the user to sign in again with `UserSession.beginOAuth2()`.

---

I hope this helps.

**Comment by btpschroeder** [comment]
Thanks @patrickarlt - wanted to acknowledge your response. I'm bringing this up with our back end folks now.

**Comment by jdelgadillo** [comment]
@patrickarlt What @btpschroeder is trying to do is basically generate this request:
`https://devext.arcgis.com/sharing/rest/oauth2/authorize?client_id=analytics4IoT&response_type=code&redirect_uri=http%3A%2F%2Fjavier.esri.com%3A9000%2Fdev`
Allowing the user to enter the username/password on ArcGIS Online and getting back the code which is then sent to backend to complete the Refresh Token Process.  We don't want the user to give our app the username/password combo if at all possible.

I've got a hacky webapp that does this in a way that our backend never sees the user's username/password combo, we're trying to replicate that functionality using the library.

**Comment by jgravois** [comment]
fwiw, you guys are not the only folks I've seen try to do this in client-side code.

the browser demo @patrickarlt shared is one of several we've put together to show how to use this library to do [Browser based named user login](https://developers.arcgis.com/documentation/core-concepts/security-and-authentication/browser-based-user-logins/).

**Comment by patrickarlt** [comment]
> We don't want the user to give our app the username/password combo if at all possible.

Both flows (client and server) allow this the entire point of Oauth is to never give usernames or passwords to third party services. If you do either flow you'll notice that the URL of the popup window (or authorize window) is always on the arcgis.com domain.

FWIW these oauth flows are how ALL ArcGIS Apps work. The only difference is that instead of a separate window sometimes the login for is displayed in an iframe which we can do because of the shared `arcgis.com` domain.

I would REALLY strongly recommend implementing one of the standard workflows based on the examples in this repo I'm happy to meet in Redlands sometime if you want.

**Comment by patrickarlt** [comment]
I'm closing this. At 4.0 `beginOAuth2` and `completeOAuth2` will use Oauth 2.1 with PKCE to get refresh token on the client.
